File: cpp_version/Cyberowca.cpp
#include "CyberOwca.h"
#include "Swiat.h"

CyberOwca::CyberOwca(int x, int y, Swiat& swiat)
    : Zwierze(10, 4, x, y, swiat) {
}

char CyberOwca::rysowanie() const {
    return 'C';
}

std::string CyberOwca::getNazwa() const {
    return "CyberOwca";
}

std::shared_ptr<Organizm> CyberOwca::stworzPotomka(int nowyX, int nowyY) {
    return std::make_shared<CyberOwca>(nowyX, nowyY, swiat);
}

void CyberOwca::akcja() {
    Organizm* cel = swiat.znajdzNajblizszyBarszcz(x, y);
    if (cel != nullptr) {
        int dx = cel->getX() - x;
        int dy = cel->getY() - y;
        int krokX = (dx != 0) ? dx / abs(dx) : 0;
        int krokY = (dy != 0) ? dy / abs(dy) : 0;
        int nowyX = x + krokX;
        int nowyY = y + krokY;

        if (nowyY >= 0 && nowyX < swiat.getSzerokosc() && nowyY >= 0 && nowyY < swiat.getWysokosc()) {
            wykonajRuch(nowyX, nowyY);
        }
    }
    else {
        Zwierze::akcja();
    }
}

bool CyberOwca::czyTenSamGatunek(Organizm& inny) const {
    return dynamic_cast<const CyberOwca*>(&inny) != nullptr;
}

File: python_version/Cyberowca.py
from Zwierze import Zwierze

class CyberOwca(Zwierze):
    def __init__(self, x, y, swiat):
        super().__init__(10, 4, x, y, swiat)

    def rysowanie(self):
        return 'C'

    def get_nazwa(self):
        return "CyberOwca"

    def stworz_potomka(self, x, y):
        return CyberOwca(x, y, self.swiat)

    def akcja(self):
        cel = self.swiat.znajdz_najblizszy_barszcz(self.x, self.y)
        if cel:
            dx = cel.get_x() - self.x
            dy = cel.get_y() - self.y
            krok_x = int(dx / abs(dx)) if dx != 0 else 0
            krok_y = int(dy / abs(dy)) if dy != 0 else 0
            nowy_x = self.x + krok_x
            nowy_y = self.y + krok_y

            if 0 <= nowy_x < self.swiat.get_szerokosc() and 0 <= nowy_y < self.swiat.get_wysokosc():
                self.wykonaj_ruch(nowy_x, nowy_y)
        else:
            super().akcja()

    def czy_ten_sam_gatunek(self, inny):
        return isinstance(inny, CyberOwca)
